neqo-transport/src/connection/mod.rs

```diff
@@ -2788,6 +2841,84 @@ impl Connection {
     pub fn stream_keep_alive(&mut self, stream_id: u64, keep: bool) -> Res<()> {
         self.streams.keep_alive(stream_id.into(), keep)
     }
+
+    fn max_dgram_size_configured(&self) -> Res<u64> {
+        let tph = self.tps.borrow();
+        if let Some(r) = &tph.remote {
+            Ok(r.get_integer(tparams::MAX_DATAGRAM_FRAME_SIZE))
+        } else if let Some(r) = &tph.remote_0rtt {
+            Ok(r.get_integer(tparams::MAX_DATAGRAM_FRAME_SIZE))
+        } else {
+            Err(Error::NotAvailable)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn set_max_dgram_size(&mut self, v: u64) {
+        self.tps
+            .borrow_mut()
+            .remote
+            .as_mut()
+            .unwrap()
+            .set_integer(tparams::MAX_DATAGRAM_FRAME_SIZE, v);
+    }
+
+    /// Returns the current max size of a datagram that can fit into a packet.
+    /// # Error
+    /// The function returns `NotAvailable` if datagrams are not enabled.
+    pub fn max_dgram_size(&self) -> Res<u64> {
+        let max_dgram_size = self.max_dgram_size_configured()?;
+        if max_dgram_size == 0 {
+            return Err(Error::NotAvailable);
+        }
+        let version = self.version();
+        let (cspace, tx) = if let Some(crypto) = self
+            .crypto
+            .states
+            .select_tx(PacketNumberSpace::ApplicationData)
+        {
+            crypto
+        } else {
+            return Err(Error::NotAvailable);
+        };
+        let path = self.paths.select_path().ok_or(Error::NotAvailable)?;
```

Comment:
IMPORTANT: You don't want to use select_path().  That's for deciding what to send next.  You want the primary path (use primary_fallible).

```suggestion
        let path = self.paths.primary_fallible().ok_or(Error::NotAvailable)?;
```